azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADOAuth2AutoConfiguration.java

```diff
@@ -0,0 +1,87 @@
+import org.springframework.security.oauth2.core.oidc.OidcIdToken;
+import org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+
+import javax.naming.ServiceUnavailableException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+
+@Configuration
+@ConditionalOnWebApplication
+@ConditionalOnProperty(prefix = "azure.activedirectory", value = {"tenant-id"})
+@EnableConfigurationProperties({AADAuthenticationProperties.class, ServiceEndpointsProperties.class})
+public class AADOAuth2AutoConfiguration {
+    private AADAuthenticationProperties aadAuthProps;
+    private ServiceEndpointsProperties serviceEndpointsProps;
+
+    public AADOAuth2AutoConfiguration(AADAuthenticationProperties aadAuthProperties,
+                                      ServiceEndpointsProperties serviceEndpointsProps) {
+        this.aadAuthProps = aadAuthProperties;
+        this.serviceEndpointsProps = serviceEndpointsProps;
+    }
+
+    @Bean
+    @ConditionalOnProperty(prefix = "azure.activedirectory", value = {"ActiveDirectoryGroups"})
+    public OAuth2UserService<OidcUserRequest, OidcUser> oidcUserService() {
+        return this.delegationOidcUserService();
+    }
+
+    private OAuth2UserService<OidcUserRequest, OidcUser> delegationOidcUserService() {
+        final OidcUserService delegate = new OidcUserService();
+
+        return (userRequest) -> {
+            // Delegate to the default implementation for loading a user
+            OidcUser oidcUser = delegate.loadUser(userRequest);
+            final OidcIdToken idToken = userRequest.getIdToken();
+
+            final String graphApiToken;
+            final Set<GrantedAuthority> mappedAuthorities;
+
+            try {
+                // https://github.com/MicrosoftDocs/azure-docs/issues/8121#issuecomment-387090099
+                // In AAD App Registration configure oauth2AllowImplicitFlow to true
+                final ClientRegistration registration = userRequest.getClientRegistration();
+                final ClientCredential credential =
+                        new ClientCredential(registration.getClientId(), registration.getClientSecret());
+
+                final AzureADGraphClient graphClient =
+                        new AzureADGraphClient(credential, aadAuthProps, serviceEndpointsProps);
+
+                graphApiToken = graphClient.acquireTokenForGraphApi(idToken.getTokenValue().toString(),
+                        aadAuthProps.getTenantId()).getAccessToken();
+
+                mappedAuthorities = graphClient.getGrantedAuthorities(graphApiToken);
+            } catch (MalformedURLException | ServiceUnavailableException
+                    | InterruptedException | ExecutionException e) {
+                throw new IllegalStateException("Failed to acquire token for Graph API.", e);
```

Comment:
The `loadUser` operation defines `throws OAuth2AuthenticationException`, so you may want to consider throwing `OAuth2AuthenticationException` here instead. Something similar to this:

```
OAuth2Error oauth2Error = new OAuth2Error("some_error_code_you_define_OR_standard_oauth_error_code");
throw new OAuth2AuthenticationException(oauth2Error, e);
```

Reply:
Thanks, will change the exception. 

@jgrandja Is there any suggested way that I can customize  or extend the pre-configured `CommonOAuth2Provider` (or `ClientRegistrationRepository` etc.), in order to add Azure Active Directory(AAD) OAuth2 endpoints as a pre-configured registration? Which allows the developer can configure AAD OAuth2 properties with less configuration, similar with `google`, `okta` etc.? Thanks in advance.

Reply:
The easiest way to provide a default set of properties in the `Environment` is as follows:

* Define a classpath resource named `aad-oauth2-common.properties` (or whatever you want to name it) with the following content:

```
spring.security.oauth2.client.provider.azure-oauth-provider.authorization-uri=http://azure-auth-server/oauth/authorize
spring.security.oauth2.client.provider.azure-oauth-provider.token-uri=http://azure-auth-server/oauth/token
spring.security.oauth2.client.provider.azure-oauth-provider.user-info-uri=http://azure-auth-server/userinfo
spring.security.oauth2.client.provider.azure-oauth-provider.jwk-set-uri=http://azure-auth-server/token_keys
spring.security.oauth2.client.provider.azure-oauth-provider.user-name-attribute=name
```

* In one of your AAD `@Configuration`, import the properties using `@PropertySource("classpath:/aad-oauth2-common.properties")`

* Now when the developer configures a client, they would specify the provider as such:

`spring.security.oauth2.client.registration.azure.provider=azure-oauth-provider`

...the client will implicitly inherit all the provider oauth endpoints so they do not need to configure this